includes/class-payment-information.php

```diff
@@ -32,31 +34,31 @@ class Payment_Information {
 	/**
 	 * Indicates whether the payment is merchant-initiated (true) or customer-initiated (false).
 	 *
-	 * @var bool
+	 * @var string
 	 */
-	private $off_session;
+	private $payment_initiated_by;
 
 	/**
 	 * Payment information constructor.
 	 *
 	 * @param string            $payment_method The ID of the payment method used for this payment.
 	 * @param \WC_Payment_Token $token The payment token used for this payment.
-	 * @param bool              $off_session Indicates whether the payment is merchant-initiated (true) or customer-initiated (false).
+	 * @param string            $payment_initiated_by Indicates whether the payment is merchant-initiated or customer-initiated.
 	 *
 	 * @throws \Exception - If no payment method is found in the provided request.
 	 */
 	public function __construct(
 		string $payment_method,
 		\WC_Payment_Token $token = null,
-		bool $off_session = false
+		string $payment_initiated_by = Payment_Initiated_By::CUSTOMER
```

Comment:
Once again throwing in `SplEnum` as an option here because if we use an enum we can guarantee that no one passes an invalid `string` to the constructor. It might feel like a bit of an overkill, but it gives us more safety during development to have the type be explicit instead of accepting any string.

I say "during development" since I don't think types are enforced at runtime (yet).

Reply:
cc @dwainm and @allendav due to previous discussions on this.

Reply:
I don't think we can use SplEnum itself, since the PECL extension might not work since we can't guarantee our merchants will have installed and using composer would cause us to run into potential collisions with any other plugin that also used such a package, no?

Or are you thinking of implementing a SplEnum like class for the plugin and using that? About what effort would be involved?

Reply:
>I don't think we can use SplEnum itself, since the PECL extension might not work since we can't guarantee our merchants will have it installed

Ah, I didn't realize the extension wasn't available by default, sorry about that!

>Or are you thinking of implementing a SplEnum like class for the plugin and using that?

Ideally, no. @DanReyLop found a php package that we can install using Composer called [php-enum](https://github.com/myclabs/php-enum) (see https://github.com/Automattic/woocommerce-payments/issues/860#issuecomment-691725813) which could work (with very little effort I think?), and it would be much better than rolling our own.

I guess at this point the question sort of becomes "is it worth it to add the additional overhead of an additional package for this sort of type safety?" and I'm a bit on the fence about that. I would love to have the additional type safety, but any solution we go with will seem like a bit of a hack.

<hr />

I guess we could also create a more complicated class with a constructor that throws an exception if you don't give it one of the correct strings and pass an object of that class to the `Payment_Information` constructor like so:

```php
class Payment_Initiated_By {
    const MERCHANT = 'initiated_by_merchant';
    const CUSTOMER = 'initiated_by_customer';

    function __construct( string $initated_by ) {
        if ( self::MERCHANT !== $initiated_by && self::CUSTOMER !== $initiated_by ) {
            throw Exception();
        }
        $this->value = $initiated_by; // and then we use this value.
    }
};

// ...

$payment_information = new Payment_Information( /* ... */ , new Payment_Initiated_By( Payment_Initiated_By::MERCHANT ) ); // OK
$payment_information = new Payment_Information( /* ... */ , new Payment_Initiated_By( 'initiated_by_merchant' ) ); // OK
$payment_information = new Payment_Information( /* ... */ , new Payment_Initiated_By( 'something_wrong' ) ); // Throws.
```

But this feels even more hacky than any of the enum solutions, so I don't really want to go down this route.

Reply:
> I guess we could also create a more complicated class with a constructor that throws an exception
Not in favor of this approach.


> ...found a php package that we can install using Composer called php-enum

Using this package feels experimental, but it would get us very close to actual type safety. Seems like it is very well maintained but even if support is lost, we could still fork it and continue using it, as the base class is really easy to understand.

The only two caveats being that 
1. we'd have to rely on magic method calls, `Payment_Initiated_By:: CUSTOMER()` instead of `Payment_Initiated_By:: CUSTOMER`.
2. When requiring a certain constant type we'd need to unwrap the value in the `Payment_Information` constructor, but again this gives us "type" surety so the trade-off seems fine.

I really like the simplicity of the implementation and would fine with the approach, lets give it a shot @reykjalin .

Reply:
I am game too, provided we are certain that if another plugin is using this there wouldn't be a conflict?

Reply:
@DanReyLop do you know if there are any conflicts we need to worry about with regards to how composer (and other plugins) handles additional packages?

Reply:
> do you know if there are any conflicts we need to worry about with regards to how composer (and other plugins) handles additional packages?

The Jetpack autoloader should take care of it. We aren't explicitly calling `require 'vendor/php-enum/...'`, so there's no risk that PHP will crash because a class is being redefined.

Worst-case-scenario, another plugin uses a newer version of `php-enum` so we accidentally include that one. But I can't imagine that the next version of `php-enum` will have breaking changes.